### DISCUSSION TITLE
Theme Switcher Idea 🎨

### DISCUSSION THREAD

**Discussion by andrewberty** [wezterm/wezterm#5144] [Ideas]
## I Implemented a theme switcher 🎨

Using input selector to allow me choose from builtin colorschemes and fuzzy search as well. and upon choosing a scheme an action will execute `sed` shell command to replace the line responsible of colorscheme in my config.

thought about sharing it maybe someone would like to tweak it and use it

Now, thinking about a way to execute the sed command whenever I move throught the colorschemes, that would provide a great live preview.

![yBDF6Lhina](https://github.com/wez/wezterm/assets/95014217/4f77a5c0-2fcb-4edd-a07d-babca4699c2d)

```lua
local wezterm = require("wezterm")
local act = wezterm.action

local M = {}

M.theme_switcher = function(window, pane)
  -- get builtin color schemes
  local schemes = wezterm.get_builtin_color_schemes()
  local choices = {}
  local config_path = "/mnt/c/Users/Andrew/.config/wezterm/wezterm.lua"

  -- populate theme names in choices list
  for key, _ in pairs(schemes) do
	table.insert(choices, { label = tostring(key) })
  end

  -- sort choices list
  table.sort(choices, function(c1, c2)
	return c1.label < c2.label
  end)

  window:perform_action(
	act.InputSelector({
	  title = "🎨 Pick a Theme!",
	  choices = choices,
	  fuzzy = true,

	  -- execute 'sed' shell command to replace the line 
          -- responsible of colorscheme in my config
	  action = wezterm.action_callback(function(inner_window, inner_pane, _, label)
		inner_window:perform_action(
		  act.SpawnCommandInNewTab({
			args = {
			  "sed",
			  "-i",
			  '/^Colorscheme/c\\Colorscheme = "' .. label .. '"',
			  config_path,
            },
		  }),
		  inner_pane
		  )
	  end),
	  }),
	  pane
  )
end

return M
```

**Comment by ben-p-commits** [comment]
@andrewberty  - I know probably want to keep this more wezterm native, but `fzf` would be able to achieve what you want.

**Reply by andrewberty** [reply]
a great idea actually, Can I run specific commands when I navigate with arrows through the list?

**Comment by ahmedelgabri** [comment]
@andrewberty thanks for this it was really helpful, one minor improvement, is that you can override the config and remove the `sed` command completely 

```lua
window:perform_action(
	act.InputSelector {
		title = '🎨 Pick a Theme!',
		choices = choices,
		fuzzy = true,

		action = wezterm.action_callback(function(inner_window, _, id, label)
			if not id and not label then
				wezterm.log_info 'cancelled'
			else
				inner_window:set_config_overrides { color_scheme = label }
			end
		end),
	},
	pane
)
```

I'm still looking for a couple of things:

- Is there a way to change the theme on select, rather than confirm. Like when I move over the choices. I couldn't find a way to do this nor an event on select for example.
- I would add a first option called "reset" to reset to the original theme.

**Reply by andrewberty** [reply]
Hi @ahmedelgabri, Thank you so much for your interest in my implementation.

Here's what I was thinking about. 

Wezterm overrides are temporary overrides, let's say you changed your colorscheme. once you close this window, wezterm has no place to memorize that.

As a lunatic, OCD, nerd myself, I always tinker with my dotfiles almost everyday. So I wanted a way to write my changes to my config file on the fly so I don't change them manually. So I thought of sed. And oh boy, I got into many troubles with sed

As this discussion is a couple of months old, I now have a different setup.

I'm using some parsers Wez has suggested to me in another discussion, which are the [serde](https://wezfurlong.org/wezterm/config/lua/wezterm.serde/index.html) parsers, basically they are parsers to/from lua and some other config languages like json and toml. So I have a `globals.toml` file which has some values that I want to change on the fly, which looks something like this

```toml
OLED = false
colorscheme = "rose-pine-moon"
opacity = 1

[font]
family = "CommitMono"

[padding]
bottom = 20
left = 20
right = 20
top = 20

```

And whenever I want to change any of them, I query that toml file parse it into lua, induce a change in that lua table, parse it back to toml and re write the whole toml file, So I don't get into troubles like I had with sed when I tried to target specific line in a huge file.

I really enjoy that approach much more. you can take a look at my wezterm dotfiles [here](https://github.com/andrewberty/dotfiles/tree/main/wezterm) If you want.

To answer your questions.
 - For implementation of live preview, I don't think there is a direct way to do so. wezterm-native, at least. I just got a dumb idea couple of days ago but didn't have time to play with it. here how it goes. what if we listened for `up` `down` keyboard events only once the input selector window is on, can we invoke an update function on every movement. I don't know, will have to try it.
 - For the reset option, a pretty good Idea actually, I just didn't think I would want to revert to wezterm original colorscheme. but you can extend the choices table with a custom choice for the default theme If that's what you want.

OFF-TOPIC NOTE!
I really love your over-engineered (as you always claim) neovim config btw. And it Inspired me in many parts of my config. So thank you so much.

I also wrote a theme switcher telescope extension [plugin](https://github.com/andrewberty/telescope-themes) for neovim, To pursue my colorscheme obsession. maybe you can also give it a try and let me know what you think. 🙌🏼

**Reply by ahmedelgabri** [reply]
The `toml` approach is an interesting one, I might have a look at it too!



> OFF-TOPIC NOTE! I really love your over-engineered (as you always claim) neovim config btw. And it Inspired me in many parts of my config. So thank you so much.
> 
> I also wrote a theme switcher telescope extension [plugin](https://github.com/andrewberty/telescope-themes) for neovim, To pursue my colorscheme obsession. maybe you can also give it a try and let me know what you think. 🙌🏼

Thanks! glad you found it useful and managed to find your way through that messy config 😄 

I will also have a look at the telescope plugin, but I'm an fzf guy because telescope is very slow in the repos I work in. So I might build my own fzf command to do the same. Thanks again

**Comment by andrewberty** [comment]
Thanks to @ahmedelgabri for bringing up fzf, I re-implemented my switchers using fzf and lua scripts to provide live-preview for whatever switcher you need

### Preview

**Colorscheme Switcher**

https://github.com/user-attachments/assets/00e61c19-f5c8-4c52-b1d5-d166a9a24a6b


**Fonts Switcher**

https://github.com/user-attachments/assets/a672eea0-afff-4264-a4d9-f976618d602b


### Example

`wezterm/features.lua`
```lua
-- fzf switcher which opens in a right split and executes a command
M.fzfSwitcher = function(window, pane, script, command)
	-- execute fzf with update script on every cursor move
	local fzfCommand = "fzf --color=gutter:-1,bg+:-1 --reverse  --preview-window=down,1 --preview='" .. script .. " {}'"

	window:perform_action(
		act.SplitPane({
			direction = "Right",
			command = {
				args = {
					"zsh",
					"-c",
					command .. fzfCommand,
				},
			},
			size = { Percent = 25 },
		}),
		pane
	)
end

M.font_switcher = function(window, pane)
	-- get system fonts by family name including only monospaced fonts, format and sort them
	local fontsCommand = "fc-list :spacing=100 family | grep -v '^\\.' | cut -d ',' -f1 | sort -u | "

	M.fzfSwitcher(window, pane, M.scriptsPath .. "/updateFont.lua", fontsCommand)
end
```
`wezterm/scripts/updateFont.lua`
```lua
#!/opt/homebrew/bin/lua

local u = dofile(os.getenv("HOME") .. "/dotfiles/wezterm/utils.lua")

local lua = u.readLuaObject(u.globalsPath)
lua.font = tostring(arg[1])
u.writeLuaObject(u.globalsPath, lua)

print(lua.font)
```

`wezterm/globals.lua`
```lua
return {
  opacity = 1,
  colorscheme = "catppuccin-mocha",
  padding = {
    bottom = 20,
    top = 20,
    left = 20,
    right = 20,
  },
  OLED = false,
  font = "Rec Mono Linear",
}
```

It is still a little bit slow as it runs the script on each cursor move and lacks reverting to the last chosen value on quit if no selection is made, but this is easy to be implemented too

Refer to my [dotfiles](https://github.com/andrewberty/dotfiles/tree/main/wezterm) for the full config
